### PR TITLE
[otbn/sw] Add ECDSA P-256 sign and verify test

### LINKED ISSUE #5140: [otbn] Create end-to-end test for ECC p256
This would be the same shape as our existing RSA test, and should allow us to estimate ECC performance numbers.

> estimate 8
> remaining 2021-02-15 2
> remaining 2021-02-22 1

### PR DESCRIPTION
**Pull request by imphil**
Add a test running on the host CPU (Ibex) to first sign a message with ECDSA using the NIST P-256 curve, and then verify it. The code matches what we have as sign and verify tests in assembly currently and has therefore the same limitations:
*  All cryto operations are done on OTBN, with the exception of the final compare step in verification (`r ==? r'`). 
* The random integer `k` is hardcoded.

Both is fine for the purpose of a first test, but we need to discuss how we want to expand this scenario to be more realistic w.r.t. a production use case. I'll open issues for that.

The used key (in PEM format) is included in the commit, as is a script that extracts key information from it and dumps it in a format suitable for the test. This is what we did for the RSA test before, and to accomodate ECC I reshuffled the script a bit.

The variable naming in the test and the assembly isn't fully consistent, with the assembly using shorter names, while the C code uses longer names (as requested previously in reviews). That's something we should make consistent as soon as the rewrite/cleanup of the P256 code is done, which @felixmiller is currently working on (see #4000).

As last commit in this series, the new test is added to FPGA CI.

Fixes #5140

### CODE REVIEW

**Comment by imphil** [comment]
This will need a rebase once #5273 is in.

@felixmiller would you mind having a look at the calling convention and the variable naming in this PR?

File: sw/device/tests/otbn/otbn_test_params.py [hunk 1]
```diff
@@ -45,16 +45,64 @@ def print_rsa_params(private_key_file: str, in_str: str) -> None:
...
+
+    print("// public key y-coordinate (Q.y)")
+    print_array("kPublicKeyQy", key.public_key().pointQ.y, key_size_bytes)
+    print("")
+
+    print("// private key (d)")
+    print_array("kPrivateKeyD", key.d, key_size_bytes)
+    print("")
+
+    print("// random integer (k)")
```
Comment (felixmiller) [inline sw/device/tests/otbn/otbn_test_params.py:81]: nit: "secret random integer"

**Review by felixmiller — APPROVED** [review]
This looks all good to me. Only nit: I would refer to `k` as "secret random integer" or "random secret".

Another minor thing ist that we currently don't supply the `rnd` value. That is used for blinding in the ECC signing algorithm. It does not matter for the result, but some intermediate values will look different, if the memory where dptr_rnd is pointing to is initialized differently.
We could add it here, to have a well-defined value for this. But this will go away anyways as soon as we have the EDN hooked up.

**Comment by imphil** [comment]
Updated:
* Use renamed assembly symbols (e.g. p256init -> p256_init)
* Consistently refer to k as "secret random integer"

**Comment by imphil** [comment]
> Another minor thing ist that we currently don't supply the rnd value.

I prefer to keep it as "undefined"/random for the time being, since we don't check the intermediate values anyways. We can then change the assembly to use the RND values as soon as that works.

File: sw/device/tests/otbn/otbn_ecdsa_p256_test.c [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+const test_config_t kTestConfig;
+
+/**
+ * CHECK()s that the actual data matches the expected data.
+ *
+ * @param actual The actual data.
+ * @param expected The expected data.
+ * @param size_bytes The size of the actual/expected data.
+ */
+static void check_data(const uint8_t *actual, const uint8_t *expected,
```
Comment (mundaym) [inline sw/device/tests/otbn/otbn_ecdsa_p256_test.c:69]: Can you add a message parameter? Just to make it quicker to see what is broken.

```c
static void check_data(const char *msg, const uint8_t *actual, const uint8_t *expected, size_t size_bytes) {
  ...
    CHECK(actual[i] == expected[i],
          "%s: mismatch at byte %d: 0x%x (actual) != 0x%x (expected)", msg, i,
          actual[i], expected[i]);
```
Call site:
```c
check_data("signature_r", signature_r, kExpectedSignatureR, 32);
```

File: sw/device/tests/otbn/otbn_ecdsa_p256_test.c [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                                 signature_r_out) == kOtbnOk);
+}
+
+/**
+ * Performs a ECDSA roundtrip test using the NIST P-256 curve.
+ *
+ * A roundtrip consists of three steps: Initialize OTBN, sign, and verify.
+ */
+static void test_ecdsa_p256_roundtrip(void) {
+  // message
```
Comment (mundaym) [inline sw/device/tests/otbn/otbn_ecdsa_p256_test.c:186]: nit: the capitalization of these comments should probably be consistent (see `// Initialize` and friends at line 225 below).

File: sw/device/tests/otbn/otbn_ecdsa_p256_test.c [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  profile_end("Initialization");
+
+  // Sign
+  uint8_t signature_r[32] = {0};
+  uint8_t signature_s[32] = {0};
+
+  LOG_INFO("Signing");
+  profile_start();
+  p256_ecdsa_sign(&otbn_ctx, kIn, kSecretRandomIntK, kPrivateKeyD, signature_r,
+                  signature_s);
```
Comment (mundaym) [inline sw/device/tests/otbn/otbn_ecdsa_p256_test.c:242]: Do we really want to include the `check_data` calls in the timing information for sign operations? If we do then should we print it separately?

**Review by mundaym — APPROVED** [review]
LGTM, just a few minor comments.

File: sw/device/tests/otbn/otbn_ecdsa_p256_test.c (hunk 4 above)
Comment (mundaym) [reply sw/device/tests/otbn/otbn_ecdsa_p256_test.c:242]: This also applies to the verify test.

File: sw/device/tests/otbn/otbn_ecdsa_p256_test.c [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+/**
+ * Ends a profiling section.
+ *
+ * The time since `profile_start()` is printed as log message.
+ *
+ * @param msg Name of the operation (for logging purposes).
+ */
+static void profile_end(const char *msg) {
+  t_end = ibex_mcycle_read();
+  uint32_t cycles = t_end - t_start;
```
[the remark this answers is not in the document]
Comment (imphil) [reply sw/device/tests/otbn/otbn_ecdsa_p256_test.c:101]: The format specifiers we use in LOG_* are special and follow more the SystemVerilog spec than the C spec, see `print.h` for details. This unfortunately means that we cannot (natively) print 64 bit values. I'll keep it therefore as uint32_t for simplicity in here, maybe we can create a proper profiling library one day (maybe that's something you could look into when you have a bit of time?) that prints something useful if we have an overflow.

File: sw/device/tests/otbn/otbn_ecdsa_p256_test.c (hunk 4 above)
Comment (imphil) [reply sw/device/tests/otbn/otbn_ecdsa_p256_test.c:242]: Sometimes :-)
* For the sign test we don't need to include it, I pulled the checking out of the profiling.
* For the verify test we want to include the checking, as it is part of the algorithm. I added a comment stating that to the code.

**Comment by imphil** [comment]
Thanks @mundaym for your useful comments! I addressed them as far as possible.

**Pull request merged**